README.md

```diff
@@ -142,9 +147,64 @@ This command will generate the codecs at the `~/git/hazelcast/custom/out`.
 
 ### Schema Validation
 
-The protocol definitions should validate against the [schema](schema/protocol-schema.json). You can configure your favorite IDE to 
-use this schema to validate and provide autocompletion.
+The protocol definitions should validate against the [schema](schema/protocol-schema.json). You can configure your IDE to 
+use this schema to validate and provide auto code-completion.
+
+The generator also uses this schema during code generation for validation purposes. It will stop and report any schema violation to the console.
+
+### Custom Types
+
+If you are going to use a custom type, a complex type that is not defined in the [currently supported types](binary/__init__.py),  
+as the type of your parameters in the protocol definitions, you need to define how to encode and decode this in the protocol level.
+
+A custom type definitions has the following structure:
+
+```yaml
+customTypes:
+    - name: CustomType1
+      since: 2.0
+      returnWithFactory: true # optional
+      params:
+        - name: paramName1
+          type: boolean
+          nullable: false
+          since: 2.0
+        - name: paramName2
+          type: String
+          nullable: true
+          since: 2.0
+```
+
+With this definition, code generator will generate a custom codec for your type and 
+calls its encode or decode methods when encoding and decoding the parameters with this custom type. 
+There are a few points to consider. 
+
+The codec for the custom type accesses the parameters defined in `params` using a 
+predefined getter pattern in its encode method. These patterns are specific to each `LANGUAGE`.
+
+For example, for the `java`, if the parameter is a `boolean` it is accessed as `customType1.isParamName1()`.
+For other types `customType1.getParamName2()` pattern is used. So, make sure that your custom types satisfies 
+this getter contract.
+
+For the decode method of the custom type codec, there are two ways to generate a
+an instance of the custom type. Default way is constructing the object using a constructor
+with parameters defined in the `params` in the order of their definition. For example, by default
+the instance of the `CustomType1` will be created with the following expression `new CustomType1(paramName1, paramName2)`.
```

Comment:
... the instance of `CustomType1` is created with the `new CustomType1(paramName1, paramName2)` expression.